File: 0786-k-th-smallest-prime-fraction/0786-k-th-smallest-prime-fraction.cpp
class Solution {
public:
    vector<int> kthSmallestPrimeFraction(vector<int>& arr, int k) {
        vector<int>ans;
        int n = arr.size();
        sort(arr.begin(), arr.end());
        priority_queue<pair<double, pair<int,int>>, vector<pair<double, pair<int,int>>>, greater<pair<double, pair<int,int>>>>pq;
        for(int i=0; i<n; i++) {
            pq.push({(double) arr[0]/arr[i], {0, i}});
        }
        while(k-- > 1 && !pq.empty()) {
            auto it = pq.top(); pq.pop();
            int x = it.second.first, y = it.second.second;
            if(x + 1 < n) {
                pq.push({(double) arr[x+1]/arr[y], {x+1, y}});
            }
        }
        if(pq.empty()) return {-1, -1};
        auto it = pq.top();
        return {arr[it.second.first], arr[it.second.second]};
    }
};

File: 0786-k-th-smallest-prime-fraction/0786-k-th-smallest-prime-fraction.java
class Solution {
    public int[] kthSmallestPrimeFraction(int[] arr, int k) {
        PriorityQueue<Pair>pq = new PriorityQueue<Pair>();
        for(int i=0; i<arr.length; i++) {
            pq.add(new Pair((double)arr[0]/arr[i], 0, i));
        }
        while(k-- > 1) {
            Pair cur = pq.poll();
            if(cur.x + 1 < arr.length) {
                pq.add(new Pair((double)arr[cur.x+1]/arr[cur.y], cur.x+1, cur.y));
            }
        }
        Pair cur = pq.poll();
        return new int[] {arr[cur.x], arr[cur.y]};
    }
    static class Pair implements Comparable<Pair> {
        double value;
        int x, y;
        Pair(double value, int x, int y) {
            this.value = value;
            this.x = x;
            this.y = y;
        }
        public int compareTo(Pair other) {
            return Double.compare(this.value, other.value);
        }
    }
}